addons/account/models/account_move.py

```diff
@@ -1609,6 +1612,88 @@ def _compute_linked_attachment_id(self, attachment_field, binary_field):
+                             LAG(other.invoice_date) OVER invoice - other.invoice_date AS date_diff
+                        FROM account_move this
+                        JOIN account_move other USING (partner_id, move_type, company_id, currency_id)
+                       WHERE other.state = 'posted'
+                         AND other.invoice_date <= COALESCE(this.invoice_date, this.date, %s)
+                         AND this.id = ANY(%s)
+                         AND this.id != other.id
+                      WINDOW invoice AS (PARTITION BY this.id ORDER BY other.invoice_date DESC)
+                ), stats AS (
+                      SELECT id,
+                             MAX(invoice_date)          OVER invoice AS last_invoice_date,
+                             AVG(date_diff)             OVER invoice AS date_diff_mean,
+                             STDDEV_SAMP(date_diff)     OVER invoice AS date_diff_deviation,
+                             AVG(amount_total)          OVER invoice AS amount_mean,
+                             STDDEV_SAMP(amount_total)  OVER invoice AS amount_deviation,
+                             ROW_NUMBER()               OVER invoice AS row_number
+                        FROM previous_invoices
+                      WINDOW invoice AS (PARTITION BY id ORDER BY invoice_date DESC)
+                )
+                  SELECT id, last_invoice_date, date_diff_mean, date_diff_deviation, amount_mean, amount_deviation
+                    FROM stats
+                   WHERE row_number BETWEEN 10 AND 30
+                ORDER BY row_number ASC
+            """, [today, draft_invoices.ids])
+            result = {invoice: vals for invoice, *vals in self.env.cr.fetchall()}
+            for move in draft_invoices:
+                invoice_date = move.invoice_date or fields.Date.context_today(self)
+                (
+                    last_invoice_date, date_diff_mean, date_diff_deviation,
+                    amount_mean, amount_deviation,
+                ) = result.get(move._origin.id, (invoice_date, 0, 10000000000, 0, 10000000000))
+
+                if date_diff_mean > 25:
+                    date_diff_deviation += 1  # correct for varying days per month and leap years
+
+                move.abnormal_date_warning = (
+                    not move.partner_id.ignore_abnormal_invoice_date
+                    and (invoice_date - last_invoice_date).days < int(date_diff_mean - 2*date_diff_deviation)
+                ) and _(
+                    "The billing frequency for %s appears unusual. Based on our historical data, the expected next invoice date is not before %s (every %s (± %s) days).\n"
+                    "Please verify if this date is accurate.",
+                    move.partner_id.display_name,
+                    format_date(self.env, fields.Date.add(last_invoice_date, days=int(date_diff_mean - 2*date_diff_deviation))),
+                    int(date_diff_mean), int(2*date_diff_deviation),
+                )
```

Comment:
> should say that the invoice might be a fake one because, to me, it's meanly an anti-scam feature more than an error detection feature

AVW detected a genuine error in his database while testing the feature, but didn't detect scam.
It avoids double encoding or other human errors that are not necessarily scam

Reply:
Hence the "or".
I'm fine if we don't add that. I just felt like it was a good idea.